src/shared/azureagent/credentials.go

```diff
@@ -0,0 +1,79 @@
+package azureagent
+
+import (
+	"context"
+	"github.com/otterize/intents-operator/src/shared/errors"
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+)
+
+const (
+	// AzureApplyOnPodLabel is used to mark pods that should be processed by the Azure agent to create an associated Azure identity & role assignment
+	AzureApplyOnPodLabel = "credentials-operator.otterize.com/create-azure-role-assignment"
+
+	// ServiceManagedByAzureAgentLabel is used to mark service accounts that are managed by the Azure agent
+	ServiceManagedByAzureAgentLabel = "credentials-operator.otterize.com/managed-by-azure-agent"
+
+	// AzureUseWorkloadIdentityLabel is used by the azure workload identity mechanism to mark pods that should use workload identity
+	AzureUseWorkloadIdentityLabel = "azure.workload.identity/use"
+
+	// AzureWorkloadIdentityClientIdAnnotation is used by the azure workload identity mechanism to link between service accounts and user assigned identities
+	AzureWorkloadIdentityClientIdAnnotation = "azure.workload.identity/client-id"
+)
+
+func (a *Agent) AppliesOnPod(pod *corev1.Pod) bool {
+	return pod.Labels != nil && pod.Labels[AzureApplyOnPodLabel] == "true"
+}
+
+func (a *Agent) OnPodAdmission(ctx context.Context, pod *corev1.Pod, serviceAccount *corev1.ServiceAccount) (updated bool, err error) {
+	logger := logrus.WithFields(logrus.Fields{"serviceAccount": serviceAccount.Name, "namespace": serviceAccount.Namespace})
+
+	if !a.AppliesOnPod(pod) {
+		return false, nil
+	}
+
+	serviceAccount.Labels[ServiceManagedByAzureAgentLabel] = "true"
+
+	pod.Labels[AzureUseWorkloadIdentityLabel] = "true"
+
+	// get or create the user assigned identity, ensuring the identity & federated credentials are in-place
+	identity, err := a.getOrCreateUserAssignedIdentity(ctx, serviceAccount.Namespace, serviceAccount.Name)
+	if err != nil {
+		return false, errors.Errorf("failed to create user assigned identity: %w", err)
+	}
+
+	clientId := *identity.Properties.ClientID
+
+	if value, ok := serviceAccount.Annotations[AzureWorkloadIdentityClientIdAnnotation]; ok && value == clientId {
```

Comment:
```suggestion
	if serviceAccount.Annotations[AzureWorkloadIdentityClientIdAnnotation] == clientId {
```

Reply:
And maybe we need to check that annotations != nil?

Reply:
we make sure that the annotations are not nill in the caller (it's kinda ugly, but saves a lot of annoyance in the agents code)